File: code/python/23.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        que,curs = [],[]
        for index,node in enumerate(lists):
            if node is not None:
                heapq.heappush(que,(node.val,index))
            curs.append(node)

        head = ListNode(-1)
        current = head
        while que:
            num,index = heapq.heappop(que)
            current.next = curs[index]
            curs[index] = curs[index].next
            current = current.next
            if curs[index] is not None:
                heapq.heappush(que,(curs[index].val,index))
        return head.next

File: code/java/23.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode head = new ListNode(-1);
        ListNode res = head;
        PriorityQueue<ListNode> minHeap = new PriorityQueue<ListNode>(new Comparator<ListNode>(){
            public int compare(ListNode l1,ListNode l2){
                return l1.val-l2.val;
            }
        });
        for(ListNode node:lists){
            if(node!=null) minHeap.offer(node);
        }
        while(!minHeap.isEmpty()){
            ListNode cur = minHeap.poll();
            ListNode temp = cur.next;
            cur.next = null;

            if(temp!=null) minHeap.offer(temp);
            head.next = cur;
            head = head.next;
        }
        return res.next;
    }
}